<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Martin Barron" />

<meta name="date" content="2017-09-17" />

<title>Sparse Differential Clustering</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Sparse Differential Clustering</h1>
<h4 class="author"><em>Martin Barron</em></h4>
<h4 class="date"><em>2017-09-17</em></h4>



<p>The SparseDC package implements the sparse differential clustering algorithm described in “A sparse differential clustering algorithm for tracing cell type changes via single-cell RNA-Sequencing data”. This algorithm clusters samples (cells) from two conditions (identifying cell types), links the clusters across conditions and identifies variables (genes) which are markers for these changes.This vignette will guide you through the steps neccessary to apply the algorithm to your data.</p>
<div id="section-1---preliminaries" class="section level2">
<h2>Section 1 - Preliminaries</h2>
<p>SparseDC takes as input data which is drawn from two conditions, for single-cell RNA-Sequencing (scRNA-seq) data, an example would be cells from populations before and after treatment. SparseDC assumes the data have been properly normalized beforehand.</p>
<p>The first step in the analysis is to load the sparseDC package:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SparseDC)
<span class="kw">set.seed</span>(<span class="dv">10</span>)</code></pre></div>
<div id="real-data-example---biase-data" class="section level3">
<h3>Real Data Example - Biase Data</h3>
<p>For this vignette we will use the scRNA-seq data created by Biase et al. to study cell fate inclination in mouse embryos <span class="citation">(Biase, Cao, and Zhong 2014)</span>. This dataset contains gene expression, FPKM, measurements for 49 cells and 16,514 genes. The cells in the dataset come from three different cell types, zygote, two-cell embryos and four-cell embryos. While the cells in this dataset are all from a single condition we have dveloped an approach to split the data into two conditions so that we can test the linking of clusters across conditions where we know the absolute truth about which clusters are present in each condition (please see the original manuscript for more details). For this dataset we put the zygote cells and half of the two-cell cells into condition 1 and the remaining two-cell cells and the four-cell cells are put into condition 2.</p>
</div>
<div id="splitting-the-data" class="section level3">
<h3>Splitting the data</h3>
<p>We first view the top of the dataset to ensure it is in the correct format. Here the genes are rows and the columns are cells as desired.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(data_biase)
<span class="kw">head</span>(data_biase[,<span class="dv">1</span>:<span class="dv">5</span>])</code></pre></div>
<pre><code>##                    GSM1377859 GSM1377860 GSM1377861 GSM1377862 GSM1377863
## ENSMUSG00000000001   3.288693   3.631147   2.290201   3.241467  3.4727581
## ENSMUSG00000000028   4.547849   4.533851   4.560077   4.682483  4.8076946
## ENSMUSG00000000037   3.662392   3.154039   3.192203   3.767524  3.2131756
## ENSMUSG00000000049   0.000000   0.000000   0.000000   0.000000  0.0000000
## ENSMUSG00000000056   2.544338   1.889089   2.143146   1.975677  1.9743810
## ENSMUSG00000000078   1.127634   1.278873   1.085679   2.132017  0.9719827</code></pre>
<p>We then store the gene names for the data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene_names_biase &lt;-<span class="st"> </span><span class="kw">row.names</span>(data_biase)</code></pre></div>
<p>We next examine the cell types of the cells present in the data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(<span class="kw">as.factor</span>(cell_type_biase))</code></pre></div>
<pre><code>## Four-cell Embryo  Two-cell Embryo           Zygote 
##               20               20                9</code></pre>
<p>With all the cell types as expected we then view the number of cells in each condition:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(<span class="kw">as.factor</span>(condition_biase))</code></pre></div>
<pre><code>##  A  B 
## 19 30</code></pre>
<p>The next step is to split the data into their respective conditions:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_A &lt;-<span class="st"> </span>data_biase[ , <span class="kw">which</span>(condition_biase ==<span class="st"> &quot;A&quot;</span>)]
data_B &lt;-<span class="st"> </span>data_biase[ , <span class="kw">which</span>(condition_biase ==<span class="st"> &quot;B&quot;</span>)]</code></pre></div>
<p>Then check that the dimensions are correct for each of the datasets:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(data_A)</code></pre></div>
<pre><code>## [1] 16514    19</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(data_B)</code></pre></div>
<pre><code>## [1] 16514    30</code></pre>
<p>And check they are in the correct format:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(data_A[ ,<span class="dv">1</span>:<span class="dv">5</span>])</code></pre></div>
<pre><code>##                    GSM1377859 GSM1377860 GSM1377861 GSM1377862 GSM1377863
## ENSMUSG00000000001   3.288693   3.631147   2.290201   3.241467  3.4727581
## ENSMUSG00000000028   4.547849   4.533851   4.560077   4.682483  4.8076946
## ENSMUSG00000000037   3.662392   3.154039   3.192203   3.767524  3.2131756
## ENSMUSG00000000049   0.000000   0.000000   0.000000   0.000000  0.0000000
## ENSMUSG00000000056   2.544338   1.889089   2.143146   1.975677  1.9743810
## ENSMUSG00000000078   1.127634   1.278873   1.085679   2.132017  0.9719827</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(data_B[ ,<span class="dv">1</span>:<span class="dv">5</span>])</code></pre></div>
<pre><code>##                    GSM1377878 GSM1377879 GSM1377880 GSM1377881 GSM1377882
## ENSMUSG00000000001  1.9891407  0.4542419   3.070172 3.15709403 4.20193210
## ENSMUSG00000000028  4.8749530  4.9535851   4.502423 4.97275342 5.23127973
## ENSMUSG00000000037  3.4309147  3.9816964   3.156383 3.02379967 2.64779051
## ENSMUSG00000000049  0.2570478  0.8972465   0.000000 0.00000000 0.17842313
## ENSMUSG00000000056  0.8593775  0.3787494   1.058426 0.06336928 0.03809113
## ENSMUSG00000000078  0.0000000  0.0000000   0.000000 0.03189917 0.00000000</code></pre>
<p>As the dimension and format of these datasets are correct we can procede to pre-processing the data and estimating the parameters for using SparseDC.</p>
</div>
<div id="section" class="section level3">
<h3></h3>
</div>
</div>
<div id="section-2---pre-processing-the-data" class="section level2">
<h2>Section 2 - Pre-processing the data</h2>
<p>SparseDC requires that data be normalized for sequencing depth and centered prior to running SparseDC. We also recommend that the data is log-transformed. To do this we have included a function that can easily pre-process the data. For the normalization it is recommended that users make use of one of the many methods that exist for normalizing scRNA-Seq data. The centering of the data is crucially important to the function of SparseDC and is vital to accurately clustering the data and identifying marker genes. We recommend that all users use this function to center their data and that only experienced users set “center=FALSE”.</p>
<p>The biase data are FPKM measurements of gene expression and so have been normalized using an alternate method as advised. This means we can set “norm = FALSE”. The biase data then needs to be both log transformed and centered so we can set “log =TRUE”&quot; and “center = TRUE”:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pre_data &lt;-<span class="st"> </span><span class="kw">pre_proc_data</span>(data_A, data_B, <span class="dt">norm =</span> <span class="ot">FALSE</span>, <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">center =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>The pre-processing function outputs the two processed datasets as a list so to extract the data we can run:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pdata_A &lt;-<span class="st"> </span>pre_data[[<span class="dv">1</span>]]
pdata_B &lt;-<span class="st"> </span>pre_data[[<span class="dv">2</span>]]</code></pre></div>
<p>And view the processed data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(pdata_A[,<span class="dv">1</span>:<span class="dv">5</span>])</code></pre></div>
<pre><code>##                      GSM1377859   GSM1377860    GSM1377861  GSM1377862
## ENSMUSG00000000001  0.162775402  0.239597984 -0.1022580244  0.15170254
## ENSMUSG00000000028 -0.001604684 -0.004130993  0.0005968605  0.02237327
## ENSMUSG00000000037  0.459405948  0.343958400  0.3531038162  0.48170453
## ENSMUSG00000000049 -0.130220364 -0.130220364 -0.1302203638 -0.13022036
## ENSMUSG00000000056  0.511663747  0.307253398  0.3915364804  0.33678380
## ENSMUSG00000000078  0.441440323  0.510111080  0.4215243542  0.82810722
##                     GSM1377863
## ENSMUSG00000000001  0.20479864
## ENSMUSG00000000028  0.04416868
## ENSMUSG00000000037  0.35809405
## ENSMUSG00000000049 -0.13022036
## ENSMUSG00000000056  0.33634818
## ENSMUSG00000000078  0.36546939</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(pdata_B[,<span class="dv">1</span>:<span class="dv">5</span>])</code></pre></div>
<pre><code>##                     GSM1377878  GSM1377879   GSM1377880  GSM1377881
## ENSMUSG00000000001 -0.19822065 -0.91872184  0.110478568  0.13160968
## ENSMUSG00000000028  0.05568304  0.06897856 -0.009826495  0.07219301
## ENSMUSG00000000037  0.40848344  0.52564787  0.344522660  0.31210403
## ENSMUSG00000000049  0.09854563  0.51018328 -0.130220364 -0.13022036
## ENSMUSG00000000056 -0.13344603 -0.43251091 -0.031746393 -0.69224534
## ENSMUSG00000000078 -0.31357011 -0.31357011 -0.313570114 -0.28216916
##                     GSM1377882
## ENSMUSG00000000001  0.35582352
## ENSMUSG00000000028  0.11456670
## ENSMUSG00000000037  0.21399903
## ENSMUSG00000000049  0.03395685
## ENSMUSG00000000056 -0.71630419
## ENSMUSG00000000078 -0.31357011</code></pre>
<p>Finally let us check the centering was succesful by examing the rowsums of the pooled data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(<span class="kw">rowSums</span>(<span class="kw">cbind</span>(pdata_A,pdata_B)))</code></pre></div>
<pre><code>##       Min.    1st Qu.     Median       Mean    3rd Qu.       Max. 
## -5.773e-15 -1.110e-15  0.000e+00 -1.578e-17  1.110e-15  5.551e-15</code></pre>
<p>As these values are all close to zero we can see that the centering has been successful and we are now ready to move on to eatimating the parameters.</p>
</div>
<div id="section-2---estimating-the-parameters" class="section level2">
<h2>Section 2 - Estimating the parameters</h2>
<p>There are two parameters to be calculated when using SparseDC, <span class="math inline">\(\lambda_{1}\)</span> and <span class="math inline">\(\lambda_{2}\)</span>, which control the level of sparsity of the marker genes for each cluster and the level of difference in marker genes for each cluster across the conditions. The estimation of these parameters is described in the original manuscript.</p>
<div id="estimating-lambda_1" class="section level3">
<h3>Estimating <span class="math inline">\(\lambda_{1}\)</span></h3>
<p>To estimate <span class="math inline">\(\lambda_{1}\)</span> all that needs to be provided is the total number of clusters present in the data and two pre-processed centered datasets. For the Biase data there are three clusters present in the dataset so we set ‘n_cluster’ equal to three.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lambda1_value &lt;-<span class="st"> </span><span class="kw">lambda1_calculator</span>(pdata_A, pdata_B, <span class="dt">ncluster =</span> <span class="dv">3</span>)
lambda1_value</code></pre></div>
<pre><code>## [1] 2.203603</code></pre>
</div>
<div id="estimating-lambda_2" class="section level3">
<h3>Estimating <span class="math inline">\(\lambda_{2}\)</span></h3>
<p>To estimate <span class="math inline">\(\lambda_{2}\)</span> we again input the centered data from the two conditions and the number of clusters in the data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lambda2_value &lt;-<span class="st"> </span><span class="kw">lambda2_calculator</span>(pdata_A, pdata_B, <span class="dt">ncluster =</span><span class="dv">3</span>)
lambda2_value</code></pre></div>
<pre><code>## [1] 0.868194</code></pre>
<p>With the <span class="math inline">\(\lambda_{1}\)</span> and <span class="math inline">\(\lambda_{2}\)</span> parameters calculated it is now time to run the SparseDC algorithm.</p>
</div>
</div>
<div id="section-3---running-sparsedc" class="section level2">
<h2>Section 3 - Running SparseDC</h2>
<p>The SparseDC function requires us to input the pre-processed centered data sets, the number of clusters and the previously calculated <span class="math inline">\(\lambda_{1}\)</span> and <span class="math inline">\(\lambda_{2}\)</span> values:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sdc_res &lt;-<span class="st"> </span><span class="kw">sparsedc_cluster</span>(pdata_A, pdata_B, <span class="dt">ncluster =</span> <span class="dv">3</span>, 
                              <span class="dt">lambda1 =</span> lambda1_value, <span class="dt">lambda2 =</span> lambda2_value)</code></pre></div>
<div id="viewing-results" class="section level3">
<h3>Viewing Results</h3>
<p>The results of SparseDC are stored as a list containing the clustering results for each condition, the cluster centers for each condition, and the scores for each of the iterations. These can be accessed by:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clusters_1 &lt;-<span class="st"> </span>sdc_res$clusters1  <span class="co"># Clusters for condition 1 data</span>
clusters_2 &lt;-<span class="st"> </span>sdc_res$clusters2  <span class="co"># Clusters for condition 2 data</span>
centers_1 &lt;-<span class="st"> </span>sdc_res$centers1  <span class="co"># Centers for condition 1 data</span>
centers_2 &lt;-<span class="st"> </span>sdc_res$centers2  <span class="co"># Centers for condition 2 data</span></code></pre></div>
<p>The results can then be seen as:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(<span class="kw">as.factor</span>(clusters_1))</code></pre></div>
<pre><code>##  1  2 
## 10  9</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(<span class="kw">as.factor</span>(clusters_2))</code></pre></div>
<pre><code>##  1  3 
## 10 20</code></pre>
<p>We can then visualize the accuracy of the clustering by comparing the clusters to the true cell types:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(cell_type_biase[<span class="kw">which</span>(condition_biase ==<span class="st"> &quot;A&quot;</span>)], clusters_1)</code></pre></div>
<pre><code>##                  clusters_1
##                    1  2
##   Two-cell Embryo 10  0
##   Zygote           0  9</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(cell_type_biase[<span class="kw">which</span>(condition_biase ==<span class="st"> &quot;B&quot;</span>)], clusters_2)</code></pre></div>
<pre><code>##                   clusters_2
##                     1  3
##   Four-cell Embryo  0 20
##   Two-cell Embryo  10  0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(<span class="kw">c</span>(cell_type_biase),<span class="kw">c</span>(clusters_1,clusters_2))</code></pre></div>
<pre><code>##                   
##                     1  2  3
##   Four-cell Embryo  0  0 20
##   Two-cell Embryo  20  0  0
##   Zygote            0  9  0</code></pre>
<p>Here see that each of the cells was clustered to the correct cell type, with the zygote cells making up cluster 1, the two-cell cells making up cluster 2 and the four-cell cells making up cluster 3.</p>
</div>
<div id="marker-genes" class="section level3">
<h3>Marker Genes</h3>
<p>The marker gene results are stored as the centers with column 1 corresponding to the marker gene results for cluster 1 and so on. Any center value not equal to zero is a marker gene for the cluster, while center values equal to zero indicate null genes. Gene with a positive center value are up-regualted for the cluster while genes with a negative center value are down-regulated for the cluster. To view the top-10 up-regulated marker genes for the zygote cell type we can run:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">zygote_top_10_index &lt;-<span class="st"> </span><span class="kw">which</span>(centers_1[,<span class="dv">1</span>] &gt;=<span class="st"> </span><span class="kw">tail</span>(<span class="kw">sort</span>(centers_1[,<span class="dv">1</span>]),<span class="dv">10</span>)[<span class="dv">1</span>])
zygote_top_10 &lt;-<span class="st"> </span>gene_names_biase[zygote_top_10_index]
zygote_top_10</code></pre></div>
<pre><code>##  [1] &quot;ENSMUSG00000052025&quot; &quot;ENSMUSG00000066772&quot; &quot;ENSMUSG00000070902&quot;
##  [4] &quot;ENSMUSG00000077323&quot; &quot;ENSMUSG00000085300&quot; &quot;ENSMUSG00000085878&quot;
##  [7] &quot;ENSMUSG00000093421&quot; &quot;ENSMUSG00000094311&quot; &quot;ENSMUSG00000096323&quot;
## [10] &quot;ENSMUSG00000097581&quot;</code></pre>
<p>To view the center values for each of the top-10 genes we can run:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">zyg_t10_res &lt;-<span class="st"> </span><span class="kw">cbind</span>(zygote_top_10, centers_1[zygote_top_10_index])
zyg_t10_res</code></pre></div>
<pre><code>##       zygote_top_10                           
##  [1,] &quot;ENSMUSG00000052025&quot; &quot;0.201965797408379&quot;
##  [2,] &quot;ENSMUSG00000066772&quot; &quot;0.119501826323416&quot;
##  [3,] &quot;ENSMUSG00000070902&quot; &quot;0.256025868166965&quot;
##  [4,] &quot;ENSMUSG00000077323&quot; &quot;0.155854111989898&quot;
##  [5,] &quot;ENSMUSG00000085300&quot; &quot;0.379620177183942&quot;
##  [6,] &quot;ENSMUSG00000085878&quot; &quot;0.203907133058937&quot;
##  [7,] &quot;ENSMUSG00000093421&quot; &quot;0.117789180618744&quot;
##  [8,] &quot;ENSMUSG00000094311&quot; &quot;0.257121704113758&quot;
##  [9,] &quot;ENSMUSG00000096323&quot; &quot;0.100485080920773&quot;
## [10,] &quot;ENSMUSG00000097581&quot; &quot;0.118753326400717&quot;</code></pre>
</div>
<div id="condition-specific-and-condition-dependent-marker-genes" class="section level3">
<h3>Condition-Specific and Condition-Dependent Marker Genes</h3>
<p>To detect condition-specific and condition-dependent marker genes (as defined in the original manuscript) we must examine the differences between the centers for clusters which are present in both conditions. As only the two-cell cells are present in each condition we examine their marker genes for condition specific and condition dependent marker genes.</p>
<p>As the two-cell cells were clustered in cluster 2 we compare the two-center vectors for cluster 2 to identify condition-specific and condition-dependent marker genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">diff_gene_index_2cell &lt;-<span class="st"> </span><span class="kw">which</span>(centers_1[,<span class="dv">2</span>] !=<span class="st"> </span>centers_2[,<span class="dv">2</span>])
diff_gene_index_2cell</code></pre></div>
<pre><code>## integer(0)</code></pre>
<p>However, for this dataset as the cells in both conditions are the same SparseDC correctly did not detect any genes as condtion-specific or condition-dependent.</p>
</div>
<div id="estimating-the-number-of-clusters-in-the-data-via-the-gap-statistic" class="section level3">
<h3>Estimating the Number of Clusters in the Data via the Gap Statistic</h3>
<p>One of the more succesful methods for estimating the number of clusters present in supervised clustering analysis is the gap statistic <span class="citation">(Tibshirani, Walther, and Hastie 2001)</span>. To aid users who may not know the number of clusters in their data prior to analysis we have included an implementation of the gap statistic for SparseDC in the R-package. Please see the original paper for full details on the method.</p>
<p>Users should beware that the gap statistic function can take some time to run especially if using a large cluster range, a high number of bootstrap samples or a high dimensional dataset. Please uncomment the code below if you would like to run it. Here “min_clus” is the minimum number of clusters to try while “max_clus” is the maximum. “nboots” controls the number of bootstrap samples used, with a default value of 200.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#gap_stat &lt;- sparsedc_gap(pdata_A, pdata_B,</span>
<span class="co">#                                 min_clus = 2, max_clus=4,</span>
<span class="co">#                                 nboots = 200, nitter = 20, nstarts = 10)</span>
<span class="co">#plot(gap_stat$gap_stat, xlab = &quot;Cluster Number&quot;, ylab = &quot;Gap Statistic&quot;,</span>
<span class="co">#     main = &quot;Gap Statistic Plot&quot;)</span>
<span class="co">#arrows(1:length(gap_stat$gap_stat),gap_stat$gap_stat-gap_stat$gap_se,</span>
<span class="co">#       1:length(gap_stat$gap_stat),gap_stat$gap_stat+gap_stat$gap_se, </span>
<span class="co">#       code=3, length=0.02, angle = 90)</span></code></pre></div>
</div>
<div id="references" class="section level3 unnumbered">
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-biase2014">
<p>Biase, Fernando H., Xiaoyi Cao, and Sheng Zhong. 2014. “Cell Fate Inclination Within 2-Cell and 4-Cell Mouse Embryos Revealed by Single-Cell RNA Sequencing.” <em>Genome Research</em> 11 (4). Genome Research: 1787–96. doi:<a href="https://doi.org/10.1101/gr.177725.114">10.1101/gr.177725.114</a>.</p>
</div>
<div id="ref-tibs2001">
<p>Tibshirani, Robert, Guenther Walther, and Trevor Hastie. 2001. “Estimating the Number of Clusters in a Data Set via the Gap Statistic.” <em>Journal of the Royal Statistical Society: Series B (Statistical Methodology)</em> 63 (2). Blackwell Publishers Ltd.: 411–23. doi:<a href="https://doi.org/10.1111/1467-9868.00293">10.1111/1467-9868.00293</a>.</p>
</div>
</div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
